libopae/plugins/xfpga/metrics/metrics.c

```diff
@@ -0,0 +1,366 @@
+	if (result != FPGA_OK) {
+		FPGA_ERR("Failed to init vector");
+		result = FPGA_INVALID_PARAM;
+		goto out_unlock;
+	}
+
+	if (objtype == FPGA_ACCELERATOR) {
+		// get AFU metrics
+		for (i = 0; i < num_metric_indexes; i++) {
+
+			result = get_afu_metric_value(handle,
+						&(_handle->fpga_enum_metric_vector),
+						metric_num[i],
+						&metrics[i]);
+			if (result != FPGA_OK) {
+				FPGA_ERR("Failed to get metric value");
+				continue;
+			}
+		}
+	} else if (objtype == FPGA_DEVICE) {
+		// get FME metrics
+		for (i = 0; i < num_metric_indexes; i++) {
+
+			result = get_fme_metric_value(handle,
+							&(_handle->fpga_enum_metric_vector),
+							metric_num[i],
+							&metrics[i]);
+			if (result != FPGA_OK) {
+				FPGA_ERR("Failed to get metric value");
+				continue;
+			}
+		}
+	} else {
+		result = FPGA_INVALID_PARAM;
+	}
+
+out_unlock:
+
+	clear_cached_values(_handle);
+	err = pthread_mutex_unlock(&_handle->lock);
+	if (err) {
+		FPGA_ERR("pthread_mutex_unlock() failed: %s", strerror(err));
+	}
+
+	return result;
+}
+
+fpga_result __FPGA_API__ xfpga_fpgaGetMetricsByName(fpga_handle handle,
+						char **metrics_names,
+						uint64_t num_metric_names,
+						struct fpga_metric *metrics)
+{
+	fpga_result result                     = FPGA_OK;
+	struct _fpga_handle *_handle           = (struct _fpga_handle *)handle;
+	int err                                = 0;
+	uint64_t i                             = 0;
+	uint64_t metric_num                    = 0;
+	fpga_objtype objtype;
+
+	if (_handle == NULL) {
+		FPGA_ERR("NULL fpga handle");
+		return FPGA_INVALID_PARAM;
+	}
+
+	result = handle_check_and_lock(_handle);
+	if (result)
+		return result;
+
+	if (_handle->fddev < 0) {
+		FPGA_ERR("Invalid handle file descriptor");
+		result = FPGA_INVALID_PARAM;
+		goto out_unlock;
+	}
+
+	if (metrics_names == NULL ||
+		metrics == NULL) {
+		FPGA_ERR("Invlaid Input parameters");
+		result = FPGA_INVALID_PARAM;
+		goto out_unlock;
+	}
+
+	if (num_metric_names == 0) {
+		FPGA_ERR("Invlaid Input parameters");
+		result = FPGA_INVALID_PARAM;
+		goto out_unlock;
+	}
+
+	result = get_fpga_object_type(handle, &objtype);
+	if (result != FPGA_OK) {
+		FPGA_ERR("Failed to init vector");
+		result = FPGA_INVALID_PARAM;
+		goto out_unlock;
+	}
+
+	if (objtype == FPGA_ACCELERATOR) {
+		// get AFU metrics
+		for (i = 0; i < num_metric_names; i++) {
+
+			result = parse_metric_num_name(metrics_names[i],
+							&(_handle->fpga_enum_metric_vector),
+							&metric_num);
+			if (result != FPGA_OK) {
+				FPGA_ERR("Invalid Input Metrics string");
+				continue;
+			}
+
+			result = get_afu_metric_value(handle, &(_handle->fpga_enum_metric_vector),
+				metric_num,
+				&metrics[i]);
+			if (result != FPGA_OK) {
+				FPGA_ERR("Failed to get metric value");
+				continue;
+			}
+
+		}
+	} else	if (objtype == FPGA_DEVICE) {
+		// get FME metrics
+		for (i = 0; i < num_metric_names; i++) {
+
+			result = parse_metric_num_name(metrics_names[i],
```

Comment:
Won't this be a relatively expensive operation? Have you done any sort of benchmark tests to see how expensive this operation is if you're polling say, every msec?